src/java/org/apache/cassandra/db/compaction/SharedCompactionProgress.java

```diff
@@ -0,0 +1,305 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.compaction;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+
+import org.apache.cassandra.io.sstable.format.SSTableReader;
+import org.apache.cassandra.schema.TableMetadata;
+
+/// Shared compaction progress tracker. This combines the progress tracking of multiple compaction tasks into a single
+/// progress tracker, and of reporting completion of all tasks when all of them complete.
+///
+/// Subtasks may start and add themselves in any order. There may also be periods of time when all started tasks have
+/// completed but there are new ones to still initiate. Because of this all parameters returned by this progress may
+/// increase over time, including the total sizes and sstable lists.
+///
+/// To know how many subtasks to expect, this class's [#registerExpectedSubtask] method must be called once per subtask
+/// before any of them start.
+public class SharedCompactionProgress implements CompactionProgress
+{
+    private final List<CompactionProgress> sources = new ArrayList<>();
+    private final AtomicInteger toComplete = new AtomicInteger(0);
+
+    public void registerExpectedSubtask()
+    {
+        toComplete.incrementAndGet();
+    }
+
+    public synchronized void addSubtask(CompactionProgress progress)
+    {
+        if (!sources.isEmpty())
+            assert sources.get(0).operationId() == progress.operationId();
+        sources.add(progress);
+    }
```

Comment:
I see that this method is synchronized, but none of the others are. How are we ensuring safe publication of the `sources` list after it is modified?

Reply:
This could cause real problems. Changed to `CopyOnWriteArrayList` and removed the synchronization here.